src/NuGet.Core/NuGet.ProjectModel/ProjectLockFile/PackagesLockFileUtilities.cs

```diff
@@ -82,153 +90,198 @@ public static bool IsLockFileStillValid(DependencyGraphSpec dgSpec, PackagesLock
-                                StringComparer.OrdinalIgnoreCase.Equals(dep.Id, p2pProjectName));
-
-                            if (projectDependency == null)
-                            {
-                                // project dependency doesn't exist in lock file.
-                                return false;
-                            }
+                    var queue = new Queue<Tuple<string, string>>();
+                    var visitedP2PReference = new HashSet<string>();
 
-                            var p2pSpec = dgSpec.GetProjectSpec(p2pUniqueName);
+                    foreach (var projectReference in framework.ProjectReferences)
+                    {
+                        if (visitedP2PReference.Add(projectReference.ProjectUniqueName))
+                        {
+                            var spec = dgSpec.GetProjectSpec(projectReference.ProjectUniqueName);
+                            queue.Enqueue(new Tuple<string, string>(spec.Name, projectReference.ProjectUniqueName));
 
-                            // The package spec not found in the dg spec. This could mean that the project does not exist anymore.
-                            if (p2pSpec != null)
+                            while (queue.Count > 0)
                             {
-                                TargetFrameworkInformation p2pSpecTargetFrameworkInformation = default;
-                                if (p2pSpec.RestoreMetadata.ProjectStyle == ProjectStyle.PackagesConfig || p2pSpec.RestoreMetadata.ProjectStyle == ProjectStyle.Unknown)
+                                var projectNames = queue.Dequeue();
+                                var p2pUniqueName = projectNames.Item2;
+                                var p2pProjectName = projectNames.Item1;
+                                var projectDependency = target.Dependencies.FirstOrDefault(
+                                    dep => dep.Type == PackageDependencyType.Project &&
+                                    StringComparer.OrdinalIgnoreCase.Equals(dep.Id, p2pProjectName));
+
+                                if (projectDependency == null)
                                 {
-                                    // Skip compat check and dependency check for non PR projects.
-                                    // Projects that are not PR do not undergo compat checks by NuGet and do not contribute anything transitively.
-                                    p2pSpecTargetFrameworkInformation = p2pSpec.TargetFrameworks.FirstOrDefault();
+                                    // new direct project dependency.
+                                    // If there are changes in the P2P2P references, they will be caught in HasP2PDependencyChanged.
+                                    invalidReasons.Add(string.Format(
+                                            CultureInfo.CurrentCulture,
+                                            Strings.PackagesLockFile_ProjectReferenceAdded,
```

Comment:
How does the message work when the proj ref added is transitive? 

Start with 3 projects: 

A -> B -> C

A has a lock file. 
Add a 4th project, A -> B -> C
and A -> B -> D, meaning B gets the extra reference.

Reply:
I have a test that for a similar scenario where the project has `A -> B -> D` P2P reference but lock file has `A -> B -> C` P2P reference. The invalid reasons for this case are displayed in the below image.

![image](https://user-images.githubusercontent.com/52756182/105902331-585a8e80-5fd3-11eb-96aa-7cd4ae73ef45.png)

Reply:
I see. This reads alright, does it still read fine when it's 2-3 levels down? 
In particular, if you need to go 2-3 levels down do you have enough info to know what happened?